src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2023,22 +2025,35 @@ private static void UpdateFormBindingArgumentExpressions(RequestDelegateFactoryC
         for (var i = 0; i < factoryContext.ArgumentExpressions.Length; i++)
         {
             var parameter = factoryContext.Parameters[i];
-            var key = parameter.Name!;
+            var formAttribute = parameter.GetCustomAttributes().OfType<IFromFormMetadata>().FirstOrDefault();
+            var key = formAttribute == null || string.IsNullOrEmpty(formAttribute.Name) ? parameter.Name! : formAttribute.Name;
```

Comment:
Ooh, good catch. Test?

Reply:
https://github.com/dotnet/aspnetcore/blob/21c81d317c543b74c0974297cfdf0abf0c95b6d2/src/Http/Http.Extensions/test/RequestDelegateFactoryTests.FormMapping.cs#L62 covers this.

This bug didn't affect previous iterations because we were always re-computing the parameter binding logic. Now that we delay it until after the metadata has been inferred, it has to match what is pre-computed in the `CreateArgument` clause.